src/domains/ValuesDomain.js

```diff
@@ -42,6 +68,187 @@ export default class ValuesDomain {
     return this._elements;
   }
 
+  // return a set of values that may be result of performing the given operation on each pair in the
+  // Cartesian product of the value sets of the operands.
+  static binaryOp(realm: Realm, op: BabelBinaryOperator, left: ValuesDomain, right: ValuesDomain) {
+    let leftElements = left._elements;
+    let rightElements = right._elements;
+    if (!leftElements || !rightElements || leftElements.size > 100 || rightElements.size > 100)
+      return ValuesDomain.topVal;
+    let resultSet = new Set();
+    try {
+      for (let leftElem of leftElements) {
+        for (let rightElem of rightElements) {
+          let result = ValuesDomain.computeBinary(realm, op, leftElem, rightElem);
+          invariant(result instanceof ConcreteValue);
+          resultSet.add(result);
+        }
+      }
+    } catch (e) {
+      if (e instanceof AbruptCompletion) return ValuesDomain.topVal;
+    }
+    return new ValuesDomain(resultSet);
+  }
+
+  static computeBinary(
```

Comment:
I tried to just keep this where it was in BinaryExpression, but that causes the Flow cycle length to increase. Arguably, this is the better place for it anyway.